### PR TITLE
Refactor compile options

### PR DESCRIPTION
**Pull request by aykevl**
This is a large refactoring PR, split into multiple pieces to make it easier to review. The first three commits just move things around, the last commit actually changes anything.

It should not affect any build. If it does, that's a bug.

### CODE REVIEW

File: compileopts/config.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/tinygo-org/tinygo/goenv"
+)
+
+// Config keeps all configuration affecting the build in a single struct.
+type Config struct {
+	Options        *Options
+	Target         *TargetSpec
```
Comment (jadr2ddude) [inline compileopts/config.go]: Is there a reason that there is a pointer here?

Comment (aykevl) [reply compileopts/config.go]: You mean `*TargetSpec`? It's used as a pointer elsewhere, so using it as a pointer here as well.

**Comment by aykevl** [comment]
I verified that none of the smoke tests had a different output. All the `md5sum`s are the same.

**Comment by aykevl** [comment]
Rebased to fix merge conflicts.

**Comment by aykevl** [comment]
Oops wrong button.

**Comment by deadprogram** [comment]
Took a while to review as it is indeed a large refactoring PR. The separation of concerns seems worthwhile. I will resist the urge to bikeshed any package naming change and instead just merge. Thanks @aykevl

**Comment by aykevl** [comment]
I wasn't quite sure about the package name either, but I couldn't think of a better name.

**Pull request merged**